### PR TITLE
[2.8] Release openstack machine before attempting to start a new instance after a timeout

### PR DESCRIPTION
**Pull request by achilleasa**
Prior to this change, if the openstack provider got a timeout while waiting for a machine to finish building it would not terminate the instance before retrying the StartInstance attempt.

This PR ensures that those machines get terminated before `StartInstance` returns with an error.

## QA steps
As this issue manifests under high load you will need to try deploying something to a loaded openstack instance.

## Bug reference
https://bugs.launchpad.net/juju/+bug/1914829

### CODE REVIEW

File: provider/openstack/provider.go [hunk 1]
```diff
@@ -1233,6 +1233,12 @@ func (e *Environ) startInstance(
 			var serverDetail *nova.ServerDetail
 			serverDetail, err = waitForActiveServerDetails(client, server.Id, 5*time.Minute)
 			if err != nil || serverDetail == nil {
+				// If we got an error back (eg. StillBuilding)
+				// we need to terminate the instance before
+				// retrying to avoid leaking resources.
+				if termErr := e.terminateInstances(ctx, []instance.Id{instance.Id(server.Id)}); termErr != nil {
+					logger.Debugf("Failed to delete instance in BUILD state, %q", termErr)
```
Comment (achilleasa) [inline provider/openstack/provider.go:1240]: This is best effort so we just log the termination error (if any) and allow the original error (`err`) to be returned by this function.

**Comment by achilleasa** [comment]
!!build!! (timeout starting build)

File: provider/openstack/provider.go [hunk 2]
```diff
@@ -1233,6 +1233,12 @@ func (e *Environ) startInstance(
 			var serverDetail *nova.ServerDetail
 			serverDetail, err = waitForActiveServerDetails(client, server.Id, 5*time.Minute)
 			if err != nil || serverDetail == nil {
+				// If we got an error back (eg. StillBuilding)
+				// we need to terminate the instance before
+				// retrying to avoid leaking resources.
+				if termErr := e.terminateInstances(ctx, []instance.Id{instance.Id(server.Id)}); termErr != nil {
```
Comment (hmlanigan) [inline provider/openstack/provider.go:1239]: Suggest adding an Info to day we're deleting the instance, like ln 1255.

File: provider/openstack/provider.go (hunk 1 above)
Comment (hmlanigan) [inline provider/openstack/provider.go:1240]: Suggest including server.ID, to aid in cleanup on failure.

It's not necessarily true that the machine is in the BUILD state here.

**Review by hmlanigan — APPROVED** [review]
Very difficult to QA without an o7k under heavy load available.  The modified unit test does fail without the provider.go change, and passes with it.

Noted a few logging suggestions

File: provider/openstack/provider.go (hunk 1 above)
Comment (achilleasa) [reply provider/openstack/provider.go:1240]: Good idea. If you get here the machine is either in the BUILD state and we timed out or we got an API error. I will revise the message tomorrow and remove the BUILD part.

File: provider/openstack/provider.go (hunk 2 above)
Comment (hmlanigan) [reply provider/openstack/provider.go:1240]: Do any of the provider's handle that case? 

The instance is leaked, as the instance id isn't saved until after it leaves the provider.

**Comment by achilleasa** [comment]
$$merge$$

**Pull request merged**